contracts/ERC827/proposals/ERC827TokenMockAllowedCallbacks.sol

```diff
@@ -0,0 +1,220 @@
+/* solium-disable security/no-low-level-calls */
+
+pragma solidity ^0.4.24;
+
+import "../../ERC20/StandardToken.sol";
+
+
+/**
+ * @title ERC827, an extension of ERC20 token standard
+ *
+ * @dev Implementation the ERC827, following the ERC20 standard with extra
+ * methods to transfer value and data and execute calls in transfers and
+ * approvals. Uses OpenZeppelin StandardToken.
+ */
+contract ERC827TokenAllowedCallbacks is StandardToken {
+
+  /**
+   * @dev Signatures fo the allowed callback allowed between contracts
+   * Receiver => Sender => Function Signature => Allowed
+   */
+  mapping(address => mapping(address => mapping(bytes4 => bool)))
+    public allowedCallbacks;
+
+  /**
+   * @dev Modifier that check the callback that wants to be executed
+   * If a receiver allows address(0) means that it can be called by anyone
+   * If a receiver allows bytes4(0) means that any function can be called
+   */
+  modifier callbackAllowed(address to, bytes4 functionSignature) {
+    // TO DO: Get first 4 bytes from data instead of using function signature
+    require(isCallbackAllowed(msg.sender, to, functionSignature));
+    _;
+  }
+
+  /**
+   * @dev Allows a callback function to be executed to a receiver contract
+   * @param from address The address that can execute the callback
+   * @param functionSignature bytes4 The signature of the callback function
+   */
+  function allowCallback(address from, bytes4 functionSignature) public {
```

Comment:
The more I think about it, the more I like it. The general use-case from my standpoint is `allowCallback('0x', 
functionSignature)`. 

On the other side, the main problem left with this solution would be working with already existing contracts. For example the DEX platform will not be able to use the goodies of approveAndCall and transferAndCall as it has no mechanism to trigger the `allowCallback` (if it is not upgradeable contract ofc). My the soluton that @k06a proposed be an answer to this question?